File: Longest Increasing Subsequence/Longest Increasing Subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        # 经典DP问题。dp[i]=长度为i的递增子串的最大的那个数，这个数保持更新为已遍历过的数字里面的最小的。
        dp = []
        for n in nums:
            pos = bisect.bisect_left(dp, n)
            if pos==len(dp): dp.append(n)
            else: dp[pos] = n
        return len(dp)

File: Longest Increasing Subsequence/Longest Increasing Subsequence.java
import java.util.Deque;

class Solution {
    public int lengthOfLIS(int[] nums) {
        if(nums.length==0) return 0;
        int res=1;
        int[] dp = new int[nums.length];
        dp[0]=1;

        for(int i=1; i<nums.length; i++){
            int max_temp=0;
            for(int j=0; j<i;j++){
                if(nums[j]<nums[i]){
                    max_temp = Math.max(max_temp, dp[j]);
                }
            }
            dp[i] = max_temp+1;
            res = Math.max(res, dp[i]);
        }
        return res;
    }
}